### PR TITLE
Bugfix 1256: reject parallel appends to unsorted data

### LINKED ISSUE #1256: `compact_incomplete` with `append=True` should fail if the existing data is `UNSORTED` or `DESCENDING`
i.e. the following test should pass:
```
@pytest.mark.parametrize("sortedness", ("DESCENDING", "UNSORTED"))
def test_parallel_append_to_known_unsorted_data(lmdb_version_store, sortedness):
    lib = lmdb_version_store
    sym = "test_parallel_append_to_known_unsorted_data"
    last_index_date = "2024-01-01" if sortedness == "DESCENDING" else "2024-01-03"
    df_0 = pd.DataFrame(
        {"col": [1, 2, 3]},
        index=[pd.Timestamp("2024-01-04"), pd.Timestamp("2024-01-02"), pd.Timestamp(last_index_date)]
    )
    lib.write(sym, df_0)
    assert lib.get_info(sym)["sorted"] == sortedness
    df_1 = pd.DataFrame({"col": [3, 4]}, index=[pd.Timestamp("2024-01-05"), pd.Timestamp("2024-01-06")])
    lib.append(sym, df_1, incomplete=True)
    with pytest.raises(SortingException):
        lib.compact_incomplete(sym, True, False)
```
Currently no exception is thrown

### PR DESCRIPTION
**Pull request by alexowens90**
Fixes #1256

### CODE REVIEW

File: cpp/arcticdb/version/version_core.cpp [hunk 1]
```diff
@@ -821,16 +821,26 @@ void read_incompletes_to_pipeline(
     pipeline_context->total_rows_ = pipeline_context->calc_rows();
 }
 
-void check_incompletes_index_ranges_dont_overlap(const std::shared_ptr<PipelineContext>& pipeline_context) {
-    // Does nothing if the symbol is not timestamp-indexed
-    // Checks both that the index ranges of incomplete segments do not overlap with one another, and that the earliest
-    // timestamp in an incomplete segment is greater than the latest timestamp existing in the symbol in the case of a
-    // parallel append
+void check_incompletes_index_ranges_dont_overlap(const std::shared_ptr<PipelineContext>& pipeline_context,
```
Comment (vasil-pashov) [inline cpp/arcticdb/version/version_core.cpp:824]: nit (style): Passing shared_ptr by const ref is [not advised](https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/).

File: cpp/arcticdb/version/version_core.cpp [hunk 2]
```diff
@@ -821,16 +821,26 @@ void read_incompletes_to_pipeline(
...
+                                                 const std::optional<SortedValue>& previous_sorted_value) {
+    /*
+     Does nothing if the symbol is not timestamp-indexed
+     Checks:
+      - that the existing data is not known to be unsorted in the case of a parallel append
+      - that the index ranges of incomplete segments do not overlap with one another
```
Comment (vasil-pashov) [inline cpp/arcticdb/version/version_core.cpp:830]: Why is this an issue? AFAIK we allow repeating index values, so in theory we can gather all incomplete segments sort them and write them to storage. Am I missing something?

File: cpp/arcticdb/version/version_core.cpp (hunk 1 above)
Comment (alexowens90) [reply cpp/arcticdb/version/version_core.cpp:824]: Fair, it is an extremely common pattern in our codebase though

Comment (willdealtry) [reply cpp/arcticdb/version/version_core.cpp:824]: PipelineContext is going to get refactored out of existence anyway

**Review by willdealtry — APPROVED** [review]

**Pull request merged**